### PR TITLE
Tickets/dm 39012

### PR DESCRIPTION
**Pull request by wvreeven**

### CODE REVIEW

**Review by r-owen — CHANGES_REQUESTED** [review]
I am very excited about this improvement. It will be wonderful to be able to use Python to run the ATMCS, and this is a big step in that direction.

My main concern is that the CSC is a simulator constructor argument, and it seems to be intended to provide structs to save and report data. Please consider writing data classes instead (or some other custom classes).

My only other significant concerns are:
* I might have found a bug in set_instrument_port (certainly code that could be clarified, even if I originally wrote it).
* Using mutable data as default values for arguments.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 1]
```diff
@@ -70,7 +78,282 @@ def __init__(self) -> None:
...
+        # Dict of M3ExitPort (the instrument port M3 points to): tuple of:
+        # * index of self.m3_port_positions: the M3 position for this port
+        # * M3State (state of M3 axis when pointing to this port)
+        # * Rotator axis at this port, as an Axis enum,
+        #   or None if this port has no rotator.
+        self._port_info_dict = {
+            M3ExitPort.NASMYTH1: (0, M3State.NASMYTH1, Axis.NA1),
+            M3ExitPort.NASMYTH2: (1, M3State.NASMYTH2, Axis.NA2),
+            M3ExitPort.PORT3: (2, M3State.PORT3, None),
+        }
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:99]: Done

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 2]
```diff
@@ -70,7 +78,282 @@ def __init__(self) -> None:
...
+        # Name of minimum limit switch event for each axis.
+        self._min_lim_names = (
+            "elevationLimitSwitchLower",
+            "azimuthLimitSwitchCW",
+            "nasmyth1LimitSwitchCW",
+            "nasmyth2LimitSwitchCW",
+            "m3RotatorLimitSwitchCW",
+        )
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:107]: I created DM-39469 for that and added a TODO.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 3]
```diff
@@ -70,7 +78,282 @@ def __init__(self) -> None:
...
+        self.limit_overtravel = 0
+        self.actuators: list[simactuators.TrackingActuator] = []
+        # allowed position error for M3 to be considered in position (deg)
+        self.m3tolerance = 1e-5
+
+    # TODO DM-38912 Make this configurable.
+    async def configure(
+        self,
+        max_tracking_interval: float = 2.5,
+        min_commanded_position: np.ndarray = np.array(
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:191]: This is copied one on one from the original code. I also opened DM-38912 to improve this as is mentioned by several TODOs in this file.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 4]
```diff
@@ -70,7 +78,282 @@ def __init__(self) -> None:
...
+            )
+            for axis in Axis
+        ]
+        self.actuators[0].verbose = True
+
+    async def connect_callback(self, server: tcpip.OneClientServer) -> None:
+        if server.connected:
+            await self.write_evt(
+                evt_id=Event.POSITIONLIMITS,
+                minimum=self.min_commanded_position.tolist(),
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:340]: No reason at all. I modified this to `list(...)`.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 5]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
     async def set_instrument_port(
         self, sequence_id: int, **kwargs: dict[str, typing.Any]
     ) -> None:
-        # TODO DM-39012: Add simulator code.
+        if CommandKey.PORT not in kwargs or self._tracking_enabled:
+            await self.write_fail_response(sequence_id=sequence_id)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:419]: Unfortunately not. I will discuss that with Te-Wei (who designed the LabVIEW TCP/IP code) to ask why and also to ask if that can be included.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 6]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+        m3_in_position = self.m3_in_position(utils.current_tai())
+        if m3actuator.target.position == m3_port_positions and m3_in_position:
+            # already there; don't do anything
+            pass
+        else:
+            self.actuators[Axis.M3].set_target(
+                tai=utils.current_tai(), position=m3_port_positions, velocity=0
+            )
+            self._axis_enabled[Axis.NA1] = False
+            self._axis_enabled[Axis.NA2] = False
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:443]: This, too, is copied one on one from the original code. If this is a bug then it was already present in the CSC.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 7]
```diff
@@ -21,41 +21,49 @@
...
-__all__ = ["McsSimulator"]
-
 CMD_ITEMS_TO_IGNORE = frozenset({CommandKey.ID, CommandKey.VALUE})
 
 
 class McsSimulator:
     """Simulate the ATMCS system."""
 
-    def __init__(self) -> None:
+    def __init__(self, csc: ATMCSCsc) -> None:
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:47]: I removed the constructor argument and introduced dataclasses instead.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:419]: Te-Wei has a suggestion how to solve this, which involves sending a second message after the `fail` message with the reason. Unfortunately the current `ts_labview_tcp_json` project doesn't support adding a reason easily. I will discuss this with Patricio and then add it. I'll create a ticket for that in due time.

[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:419]: It's unfortunate not to be able to include all the related information in the same message, but this sounds better than nothing having the reason available.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 8]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+            self._axis_enabled[Axis.NA1] = False
+            self._axis_enabled[Axis.NA2] = False
+            await self.update_events()
         await self.write_success_response(sequence_id=sequence_id)
 
     async def start_tracking(
         self, sequence_id: int, **kwargs: dict[str, typing.Any]
     ) -> None:
-        # TODO DM-39012: Add simulator code.
+        m3_in_position = self.m3_in_position(utils.current_tai())
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:450]: Once again I am copying code from the CSC and the corresponding methods didn't have a doc string there either. Having said that, I am following up on all your suggestions.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 9]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+        self._stop_tracking_task.cancel()
+        self._stop_tracking_task = asyncio.create_task(self._finish_stop_tracking())
+        await self.update_events()
         await self.write_success_response(sequence_id=sequence_id)
 
     async def track_target(
         self, sequence_id: int, **kwargs: dict[str, typing.Any]
     ) -> None:
-        # TODO DM-39012: Add simulator code.
+        if not self._tracking_enabled:
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:487]: I have made all arguments explicit.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 10]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+        if not m3_in_position:
+            await self.write_fail_response(sequence_id=sequence_id)
+            return
+        if not self._stop_tracking_task.done():
+            await self.write_fail_response(sequence_id=sequence_id)
+            return
+        self._tracking_enabled = True
+        await self.update_events()
+        # TODO DM-39408 Not sure what to do here. The original code in the CSC
+        #  set the CSC to FAULT after a certain period was passed but there is
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:460]: I retained the original code with a new TODO indicating that a feedback mechanism from the ATMCS to the CSC needs to be implemented. Perhaps a new event (not a SAL event!!!) message would be sufficient so the CSC can take proper action. This certainly is future work.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 11]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+        # self._set_tracking_timer(restart=True)
         await self.write_success_response(sequence_id=sequence_id)
 
     async def stop_tracking(
         self, sequence_id: int, **kwargs: dict[str, typing.Any]
     ) -> None:
-        # TODO DM-39012: Add simulator code.
+        if not self._stop_tracking_task.done():
+            await self.write_fail_response(sequence_id=sequence_id)
+            return
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:470]: Again, this is copied one on one from the original code. I agree that it needs to be revised but that is future work.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py [hunk 12]
```diff
@@ -132,27 +415,619 @@ async def verify_data(self, data: dict[str, typing.Any]) -> bool:
...
+
+        # TODO DM-39408 Not sure what to do here. The original code in the CSC
+        #  set the CSC to FAULT after a certain period was passed but there is
+        #  no way to do that with the TCP/IP infrastructure.
+        # self._set_tracking_timer(restart=True)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:475]: TODO removed. See my previous comment about possibly introducing a new event message.

File: python/lsst/ts/atmcssimulator/dataclasses.py [hunk 13]
```diff
@@ -0,0 +1,401 @@
...
+        self.m3EncoderRaw = [0.0] * 100
+
+
+@dataclass
+class PortInfo:
+    """Dataclass holding port information.
+
+    Attributes
+    ----------
+    index : int
```
Comment (r-owen) [inline python/lsst/ts/atmcssimulator/dataclasses.py:309]: Please fix the documented types to match the actual type (int | None in this case)

File: python/lsst/ts/atmcssimulator/mcs_simulator.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:213]: Scary. I didn't spend a lot of time on this code, but still...

File: python/lsst/ts/atmcssimulator/mcs_simulator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:501]: Actually this is correct; I just forgot why it was done. When the tertiary mirror is moved then the NA1 and NA2 rotators are disabled.

If you are willing, please aded a code comment explaining that.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:536]: Fair enough and best left until we know what the real controller does.

File: python/lsst/ts/atmcssimulator/mcs_simulator.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (r-owen) [reply python/lsst/ts/atmcssimulator/mcs_simulator.py:501]: Actually this is correct; I just forgot why it was done. When the tertiary mirror is moved then the NA1 and NA2 rotators are disabled.

If you are willing, please aded a code comment explaining that.

**Review by r-owen — APPROVED** [review]
I have a few smaller cleanups I hope you will consider before merging, the main one being to use a different way of specifying the default for your new dataclasses (simple -- a global search/replace plus removing the `__post_init__` methods).

File: python/lsst/ts/atmcssimulator/dataclasses.py [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+        The elevation raw encoder angle [deg].
+
+    """
+
+    azimuth1Encoder: list[float] = field(default_factory=list)
+    azimuth1EncoderRaw: list[float] = field(default_factory=list)
+    azimuth2Encoder: list[float] = field(default_factory=list)
+    azimuth2EncoderRaw: list[float] = field(default_factory=list)
+    elevationEncoder: list[float] = field(default_factory=list)
+    elevationEncoderRaw: list[float] = field(default_factory=list)
```
[the remark this answers is not in the document]
Comment (wvreeven) [reply python/lsst/ts/atmcssimulator/dataclasses.py:78]: Thanks. This is a much better solution and it wasn't clear to me from the Python docs that this is possible.

**Pull request merged**